source/beamline/BioXAS/BioXASBiStateGroup.cpp

```diff
@@ -0,0 +1,363 @@
+
+BioXASBiStateGroup::BioXASBiStateGroup(const QString &name, QObject *parent) :
+	AMEnumeratedControl(name, "", parent)
+{
+
+}
+
+BioXASBiStateGroup::~BioXASBiStateGroup()
+{
+
+}
+
+bool BioXASBiStateGroup::canMeasure() const
+{
+	bool result = false;
+
+	if (isConnected()) {
+
+		// Iterate through the list of children, finding out if
+		// all can be measured.
+
+		QList<AMControl*> children = childControls();
+
+		if (children.count() > 0) {
+			bool measurable = true;
+
+			for (int i = 0, count = children.count(); i < count && measurable; i++) { // we want to stop if any one child is not measurable.
+				AMControl *child = children.at(i);
+
+				if (child && child->canMeasure())
+					measurable = true;
+				else
+					measurable = false;
+			}
+
+			result = measurable;
+		}
+	}
+
+	return result;
+}
+
+bool BioXASBiStateGroup::canMove() const
+{
+	bool result = false;
+
+	if (isConnected()) {
+
+		// Iterate through the list of children, finding out if
+		// any can be moved.
+
+		QList<AMControl*> children = childControls();
+
+		if (children.count() > 0) {
+			bool movable = false;
+
+			for (int i = 0, count = children.count(); i < count && !movable; i++) { // we want to stop if we find one child that can be moved.
+				AMControl *child = children.at(i);
+
+				if (child && child->canMove())
+					movable = true;
+				else
+					movable = false;
+			}
+
+			result = movable;
+		}
+	}
+
+	return result;
+}
+
+bool BioXASBiStateGroup::canStop() const
+{
+	bool result = false;
+
+	if (isConnected()) {
+
+		// Iterate through the list of children, finding out if
+		// those that can move can also be stopped.
+
+		QList<AMControl*> children = childControls();
+
+		if (children.count() > 0) {
+			bool childrenValid = true;
+			bool childrenStoppable = true;
+
+			for (int i = 0, count = children.count(); i < count && childrenValid && childrenStoppable; i++) { // we want to stop if we come across any invalid children or children that 'should' be stoppable but aren't.
+				AMControl *child = children.at(i);
+
+				if (child) {
+					if (child->canMove() && !child->canStop()) // we want to know if we have any children that can move but can't be stopped. Potential problem!
+						childrenStoppable = false;
+
+				} else {
+					childrenValid = false; // we want to know if we have any children that aren't valid. Another potential problem.
+				}
+			}
+
+			result = childrenValid && childrenStoppable;
+		}
+	}
+
+	return result;
+}
+
+void BioXASBiStateGroup::updateConnected()
+{
+	bool connected = false;
+
+	// Iterate through list of children, finding out if
+	// all are connected.
+
+	QList<AMControl*> children = childControls();
+
+	if (children.count() > 0) {
+
+		bool childrenConnected = true;
+
+		for (int i = 0, count = children.count(); i < count && childrenConnected; i++) { // we want to stop if any one child isn't connected.
+			AMControl *child = children.at(i);
+
+			if (child && child->isConnected())
```

Comment:
This is a little unnecessary, the value of `childrenConnected` is always true until it isn't.  This should probably just be,

```c++
if (!(child && child->connected())
     childrenConnected = false;
```